content/chapters/io/lab/content/file-operations.md

```diff
@@ -0,0 +1,46 @@
+# File Handling
+
+You've most likely had to deal with files in the past.
+Go to `support/simple-file-handling` and run a most basic command: `cat file.txt`:
+
+```console
+student@os:~/.../lab/support/simple-file-handling$ cat file.txt
+OS Rullz!
+```
+
+But how does `cat` actually "reach" to the file, then read its contents, then print them to standard output?
+This is part of what we're going to learn.
+
+## Reaching the File
+
+To manipulate the file (read its contents, modify them, change its size etc.), each process must first get a **handler** to this file.
+Think of this handler as an object by which the process can identify and refer to the file.
+
+Now take a look at the code examples in `support/simple-file-handling`.
+Each of them reads the contents of `file.txt`, modifies them and then reads the previously modified file again.
+Use `make` to compile the C code and `make java-file-operations` to compile the Java code.
+
+Now run the programs repeatedly in whatever order you wish:
+
+```console
+student@os:~/.../lab/support/simple-file-handling$ python3 file_operations.py
+File contents are: OS Rullz!
+Wrote new data to file
+File contents are: Python was here!
+
+student@os:~/.../lab/support/simple-file-handling$ ./file_operations  # from the C code
+File contents are: Python was here!
+Wrote new data to file
+File contents are: C was here!
+
+student@os:~/.../lab/support/simple-file-handling$ java FileOperations
+File contents are: Python was here!
+Wrote new data to file
+File contents are: C was here!
+```
+
+Note that each piece of code creates a variable which is then used as a handler.
+
+[Quiz](../quiz/file-handler-c.md)
+
+TODO (later in the lab, after file descriptors): handler is `open_file` struct -> link to kernel code.
```

Comment:
We could include only a snippet highlighting the main aspects: offset, flags, file_ops, and inode. 
I am thinking of something similar to what @teodutu did for [ task_struct](https://github.com/open-education-hub/operating-systems/blob/master/content/chapters/compute/lecture/slides/process-attributes.md#process-attributes---from-kernel-space-1) in Compute lecture:

```c
struct file {
	struct path		f_path;

  /* Data and metadata */
	struct inode	*f_inode;

  /* Implementation for file operations i.e. read, write */
	const struct  file_operations *f_op;

	spinlock_t		f_lock;

  /* Referrence count*/
	atomic_long_t	f_count;

  /* Open mode */
	unsigned int 	f_flags;
	fmode_t			  f_mode;

	struct mutex	f_pos_lock;
	loff_t			  f_pos;

#ifdef CONFIG_EPOLL
	struct hlist_head	*f_ep;
#endif /* #ifdef CONFIG_EPOLL */
}
```

Reply:
Yup, that's what I intended.